src/workerd/io/worker.c++

```diff
@@ -985,6 +985,14 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
   lock->setCaptureThrowsAsRejections(features.getCaptureThrowsAsRejections());
   lock->setCommonJsExportDefault(features.getExportCommonJsDefaultNamespace());
 
+  if (features.getNodeJsCompat()) {
+    // Enabling async context tracking installs a relatively expensive callback on the v8 isolate
```

Comment:
Now that we're not copying the map on every promise continuation, is it actually that expensive? Have you done any measurement of this? If the performance overhead is not large then I'd suggest we always enable it.

If there is a lot of overhead, maybe we can do a trick where we enable the hook only when the API is first used. We would then have to assume that any promise which doesn't have an async context attached must be using the root (empty) context.

(That also suggests a further optimization: maybe we shouldn't bother attaching anything to the promise when the current context is the root context.)

Reply:
We're not copying the map but we are allocating a new opaque object and setting the private property on every kInit which still adds a fair amount of overhead on its own. There's also the overhead of performing lookups on each event.

Lazily enabling the promise hook on first use is certainly an option. I think it could be a bit weird tho since it would be enabled for the entire isolate. That is, we could have a case where one request doesn't enable it (beacuse the api is not used), the next request does (because the api is used), and all subsequent uses of that isolate has it enabled. I think I'd rather just have it always on or always off. 

I do like the idea of not attaching anything if the current is the root.

Reply:
Ok, I've implemented the bit here about not attaching if current is root but left the rest here as is.

Reply:
> We're not copying the map but we are allocating a new opaque object and setting the private property on every kInit which still adds a fair amount of overhead on its own.

Oh, maybe we should improve this further.

We could avoid allocating a new opaque object if `AsyncContextFrame` inherited `Wrappable` and used the usual wrapping mechanisms -- then each frame would only need to allocate a wrapper once.

Also if we can just not attach the context to the promise at all when the promise is in the root context, then when people aren't using ALS, the overhead should be minimal. Still some overhead to call the callback, but no changes made.

In that  case, combined with not registering the hook at all until the API is first used, I think we'd be in pretty good shape and would not need a global switch.

We should, of course, actually measure the overhead. You can do this pretty easily with test-runner in --load-test mode. Run a sample test that does some things with promises for a set number of iterations before and after the change.

Reply:
With a simple variation of one of our blob tests (known test that is fairly promise dependent)... With the current variation where the promise hook is installed automatically when the nodeJsCompat flag is enabled.. the load test results are:

* Without the flag enabled: 1000 requests, ~218 RPS
* With the flag enabled: 1000 requests, ~195 RPS

Reply:
Given the perf results there I'll go ahead and lazily enable only when the ALS api is used.